From b93a2617d49d9636801130d974995cabe6335b71 Mon Sep 17 00:00:00 2001
From: Frantisek Sumsal <frantisek@sumsal.cz>
Date: Mon, 1 Jul 2019 09:27:59 +0200
Subject: [PATCH] test: replace `tail -f` with journal cursor which should
 be...

more reliable

(cherry picked from commit d0b2178f3e79f302702bd7140766eee03643f734)

Related: #1808940
---
 test/TEST-36-NUMAPOLICY/testsuite.sh | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/test/TEST-36-NUMAPOLICY/testsuite.sh b/test/TEST-36-NUMAPOLICY/testsuite.sh
index e15087b137..306a96b517 100755
--- a/test/TEST-36-NUMAPOLICY/testsuite.sh
+++ b/test/TEST-36-NUMAPOLICY/testsuite.sh
@@ -29,6 +29,9 @@ testUnitNUMAConf="$testUnitFile.d/numa.conf"
 journalSleep=5
 sleepAfterStart=1
 
+# Journal cursor for easier navigation
+journalCursorFile="jounalCursorFile"
+
 startStrace() {
     coproc strace -qq -p 1 -o $straceLog -e set_mempolicy -s 1024 $1
 }
@@ -38,18 +41,16 @@ stopStrace() {
 }
 
 startJournalctl() {
-    coproc journalctl -u init.scope -f > $journalLog
+    # Save journal's cursor for later navigation
+    journalctl --no-pager --cursor-file="$journalCursorFile" -n0 -ocat
 }
 
 stopJournalctl() {
-    # Wait a few seconds until the messages get properly queued...
-    sleep $journalSleep
-    # ...and then force journald to write them to the backing storage
-    # Also, using journalctl --sync should be better than using SIGRTMIN+1, as
+    # Using journalctl --sync should be better than using SIGRTMIN+1, as
     # the --sync wait until the synchronization is complete
     echo "Force journald to write all queued messages"
     journalctl --sync
-    kill -s TERM $COPROC_PID
+    journalctl -u init.scope --cursor-file="$journalCursorFile" > "$journalLog"
 }
 
 checkNUMA() {
